File: laundry/wash_wallet_be/resources/js/types/outlet.ts
import {
    BaseFilters,
    BaseSortOptions,
    Category,
    CourierSchedule,
    CourierSetting,
    Customer,
    Employee,
    Expense,
    Fine,
    JournalEntry,
    LaundryService,
    MembershipPlan,
    OperationalDay,
    OutletFeature,
    OutletSetting,
    Position,
    ServicePackage,
    User,
} from ".";

export interface Outlet {
    id: number;
    ownerId: number;
    name: string;
    code: string;
    email: string;
    provinceId?: number | null;
    provinceName?: string | null;
    cityId?: number | null;
    cityName?: string | null;
    districtId?: number | null;
    districtName?: string | null;
    villageId?: number | null;
    villageName?: string | null;
    street?: string | null;
    phone?: string | null;
    status: "inactive" | "trial" | "active" | "expired";
    statusLabel: string;
    coinBalance: number;
    createdAt: string;
    updatedAt: string;
    deletedAt?: string | null;
    fullAddress?: string;
    latitude?: number | null;
    longitude?: number | null;
    categories?: Category[];
    courierSchedules?: CourierSchedule[];
    courierSetting?: CourierSetting;
    customers?: Customer[];
    employees?: Employee[];
    expenses?: Expense[];
    fines?: Fine[];
    journalEntries?: JournalEntry[];
    laundryServices?: LaundryService[];
    membershipPlans?: MembershipPlan[];
    operationalDays: OperationalDay[];
    outletFeatures?: OutletFeature[];
    outletSettings?: OutletSetting[];
    positions?: Position[];
    servicePackages?: ServicePackage[];
    owner: User;
    categoriesCount?: number;
    courierSchedulesCount?: number;
    customersCount?: number;
    employeesCount?: number;
    expensesCount?: number;
    finesCount?: number;
    journalEntriesCount?: number;
    laundryServicesCount?: number;
    membershipPlansCount?: number;
    positionsCount?: number;
    ordersCount?: number;
    outletFeaturesCount?: number;
    servicePackagesCount?: number;
    isActivated?: boolean;
    isActive?: boolean;
    activationStatus?: {
        status: "inactive" | "trial" | "active" | "expired";
        trialStartedAt?: string | null;
        trialExpiresAt?: string | null;
        unlockedAt?: string | null;
        expiresAt?: string | null;
        trialRemainingDays: number;
        trialEligible?: boolean;
        trialEligibilityCode?: string | null;
        trialEligibilityMessage?: string | null;
        trialDurationDays?: number;
    } | null;
    isCourierEnabled?: boolean;
    hasFreeShipping?: boolean;
    hasUnconditionalFreeShipping?: boolean;
}

export interface OutletFilters extends BaseFilters {
    provinceId?: number;
    districtId?: number;
    cityId?: number;
    status?: "active" | "inactive" | "trial" | "expired";
}

export interface OutletSortOptions extends BaseSortOptions {
    column:
        | "name"
        | "code"
        | "email"
        | "createdAt"
        | "updatedAt"
        | "status"
        | "categoriesCount"
        | "employeesCount";
}

export interface OutletFormData {
    name: string;
    email?: string | null;
    phone?: string | null;
    provinceId?: number | null;
    provinceName?: string | null;
    cityId?: number | null;
    cityName?: string | null;
    districtId?: number | null;
    districtName?: string | null;
    villageId?: number | null;
    villageName?: string | null;
    street?: string | null;
    isActive?: boolean | null;
    latitude?: number | null;
    longitude?: number | null;
    [key: string]: any;
}

export interface OutletUploadImportFormData {
    file: File;
    type: "outlet";
    [key: string]: any;
}

export interface OutletCategoryFormData {
    name: string;
    description?: string;
    isActive?: boolean | null;
    [key: string]: any;
}

export interface OutletCategoryUploadImportFormData {
    file: File;
    type: "category";
    [key: string]: any;
}

export interface OutletCommissionFormData {
    unitId: number;
    name: string;
    description?: string;
    [key: string]: any;
}

export interface OutletCustomerFormData {
    name: string;
    email?: string;
    phone?: string;
    gender: "male" | "female" | "";
    address: string;
    isActive?: boolean | null;
    [key: string]: any;
}

export interface OutletCustomerUploadImportFormData {
    file: File;
    type?: string;
    [key: string]: any;
}

export interface EmployeeSalaryFormItem {
    salaryId: number | null;
    type: string;
    status: string;
    amount: number;
    [key: string]: any;
}

export interface EmployeeProcessCommissionFormItem {
    processId: number | null;
    commissionType: "per_item" | "per_kg" | "percentage" | "flat";
    commissionValue: number;
    hasTarget: boolean;
    targetThreshold: number | null;
    bonusAmount: number | null;
    effectiveDate?: string;
    isActive: boolean;
    [key: string]: any;
}

export interface OutletEmployeeCreateFormData {
    name: string;
    username: string;
    password?: string | null;
    passwordConfirmation?: string | null;
    avatar?: File | null;
    phone: string;
    address: string;
    gender: "male" | "female" | "";
    dateOfBirth?: string;
    startDate: string;
    isActive: boolean;
    cutoffDays: number;
    positionIds: number[];
    employeeSalaries: EmployeeSalaryFormItem[];
    employeeProcessCommissions: EmployeeProcessCommissionFormItem[];
    [key: string]: any;
}

export interface OutletEmployeeEditFormData {
    name: string;
    username: string;
    password?: string | null;
    passwordConfirmation?: string | null;
    avatar?: File | null;
    phone: string;
    address: string;
    gender: "male" | "female" | "";
    dateOfBirth?: string;
    startDate: string;
    isActive: boolean;
    cutoffDays: number;
    [key: string]: any;
}

export interface OutletFineFormData {
    name: string;
    amount: number;
    description: string;
    [key: string]: any;
}
export interface OutletLaundryServiceFormData {
    categoryId: number;
    unitId: number;
    name: string;
    description?: string;
    isActive?: boolean | null;
    supportsCourier?: boolean | null;
    price: number;
    durationHours: number;
    minQuantity: number;
    laundryServiceProcesses?: Array<{
        processId: number;
    }>;
    [key: string]: any;
}

export interface OutletMembershipPlanFormData {
    name: string;
    price: number;
    durationDays: number | null;
    discountPercentage: number | null;
    description: string;
    level: number;
    isActive?: boolean | null;
}

export interface OutletOperationalDayFormData {
    dayOfWeek: string;
    isOpen: boolean;
    openTime?: string;
    closeTime?: string;
    [key: string]: any;
}

export interface OutletPositionFormData {
    name: string;
    description?: string;
    isActive?: boolean | null;
    permissions: string[];
    [key: string]: any;
}

export interface ServicePackageItemFormData {
    laundryServiceId: number;
    quantity: number;
}

export interface OutletServicePackageFormData {
    name: string;
    description: string;
    price: number;
    validityDays: number | null;
    isActive: boolean;
    servicePackageItems: ServicePackageItemFormData[];
}

File: laundry/wash_wallet/packages/wash_wallet_domain/lib/src/entities/outlet.dart
import 'package:equatable/equatable.dart';
import 'category.dart';
import 'operational_day.dart';
import 'outlet_operational_status.dart';

class Outlet extends Equatable {
  final int id;
  final int? ownerId;
  final String name;
  final String? code;
  final String? phone;
  final String? email;
  final String? description;
  final bool isActive;
  final int coinBalance;
  final int? provinceId;
  final String? provinceName;
  final int? cityId;
  final String? cityName;
  final int? districtId;
  final String? districtName;
  final int? villageId;
  final String? villageName;
  final String? street;
  final String? fullAddress;
  final String? statusLabel;
  final String? createdAt;
  final String? updatedAt;
  final String? deletedAt;
  final double? latitude;
  final double? longitude;
  final double? distance;
  final bool hasExposure;
  final bool isActivated;
  final bool hasActiveExposure;
  final bool isExposureExpired;
  final bool isCurrentlyOpen;
  final bool isCourierEnabled;
  final bool hasFreeShipping;
  final bool hasUnconditionalFreeShipping;
  final OutletOperationalStatus? operationalStatus;
  final Map<String, dynamic>? todaySchedule;
  final Map<String, dynamic>? nextOpenDay;
  final int? categoriesCount;
  final int? customersCount;
  final int? employeesCount;
  final int? finesCount;
  final int? laundryServicesCount;
  final int? operationalDaysCount;
  final int? membershipPlansCount;
  final int? positionsCount;
  final int? ordersCount;
  final int? expensesCount;
  final int? journalEntriesCount;
  final int? servicePackagesCount;
  final double? averageRating;
  final int? reviewsCount;
  final List<Category>? categories;
  final List<OperationalDay>? operationalDays;

  const Outlet({
    required this.id,
    this.ownerId,
    required this.name,
    this.code,
    this.phone,
    this.email,
    this.description,
    required this.isActive,
    this.coinBalance = 0,
    this.provinceId,
    this.provinceName,
    this.cityId,
    this.cityName,
    this.districtId,
    this.districtName,
    this.villageId,
    this.villageName,
    this.street,
    this.fullAddress,
    this.statusLabel,
    this.createdAt,
    this.updatedAt,
    this.deletedAt,
    this.categoriesCount,
    this.customersCount,
    this.employeesCount,
    this.finesCount,
    this.laundryServicesCount,
    this.operationalDaysCount,
    this.membershipPlansCount,
    this.positionsCount,
    this.ordersCount,
    this.expensesCount,
    this.journalEntriesCount,
    this.servicePackagesCount,
    this.latitude,
    this.longitude,
    this.distance,
    this.hasExposure = false,
    this.isActivated = false,
    this.hasActiveExposure = false,
    this.isExposureExpired = false,
    this.isCurrentlyOpen = false,
    this.isCourierEnabled = true,
    this.hasFreeShipping = false,
    this.hasUnconditionalFreeShipping = false,
    this.operationalStatus,
    this.todaySchedule,
    this.nextOpenDay,
    this.averageRating,
    this.reviewsCount,
    this.categories,
    this.operationalDays,
  });

  factory Outlet.fromModel(dynamic model) {
    return Outlet(
      id: model.id,
      ownerId: model.ownerId,
      name: model.name,
      code: model.code,
      phone: model.phone,
      email: model.email,
      description: model.description,
      isActive: model.isActive,
      coinBalance: model.coinBalance,
      provinceId: model.provinceId,
      provinceName: model.provinceName,
      cityId: model.cityId,
      cityName: model.cityName,
      districtId: model.districtId,
      districtName: model.districtName,
      villageId: model.villageId,
      villageName: model.villageName,
      street: model.street,
      fullAddress: model.fullAddress,
      statusLabel: model.statusLabel,
      createdAt: model.createdAt,
      updatedAt: model.updatedAt,
      deletedAt: model.deletedAt,
      categoriesCount: model.categoriesCount,
      customersCount: model.customersCount,
      employeesCount: model.employeesCount,
      finesCount: model.finesCount,
      laundryServicesCount: model.laundryServicesCount,
      operationalDaysCount: model.operationalDaysCount,
      membershipPlansCount: model.membershipPlansCount,
      positionsCount: model.positionsCount,
      ordersCount: model.ordersCount,
      expensesCount: model.expensesCount,
      journalEntriesCount: model.journalEntriesCount,
      servicePackagesCount: model.servicePackagesCount,
      latitude: model.latitude,
      longitude: model.longitude,
      distance: model.distance,
      hasExposure: model.hasExposure,
      isActivated: model.isActivated,
      hasActiveExposure: model.hasActiveExposure,
      isExposureExpired: model.isExposureExpired,
      isCurrentlyOpen: model.isCurrentlyOpen,
      isCourierEnabled: model.isCourierEnabled,
      hasFreeShipping: model.hasFreeShipping,
      hasUnconditionalFreeShipping: model.hasUnconditionalFreeShipping,
      operationalStatus: model.operationalStatus?.toEntity(),
      todaySchedule: model.todaySchedule,
      nextOpenDay: model.nextOpenDay,
      averageRating: model.averageRating,
      reviewsCount: model.reviewsCount,
      categories: model.categories?.map((e) => e.toEntity()).toList(),
      operationalDays: model.operationalDays?.map((e) => e.toEntity()).toList(),
    );
  }

  @override
  List<Object?> get props => [
    id,
    ownerId,
    name,
    code,
    phone,
    email,
    description,
    isActive,
    coinBalance,
    provinceId,
    provinceName,
    cityId,
    cityName,
    districtId,
    districtName,
    villageId,
    villageName,
    street,
    fullAddress,
    statusLabel,
    createdAt,
    updatedAt,
    deletedAt,
    categoriesCount,
    customersCount,
    employeesCount,
    finesCount,
    laundryServicesCount,
    operationalDaysCount,
    membershipPlansCount,
    positionsCount,
    ordersCount,
    expensesCount,
    journalEntriesCount,
    servicePackagesCount,
    latitude,
    longitude,
    distance,
    hasExposure,
    isActivated,
    hasActiveExposure,
    isExposureExpired,
    isCurrentlyOpen,
    isCourierEnabled,
    hasUnconditionalFreeShipping,
    hasFreeShipping,
    operationalStatus,
    todaySchedule,
    nextOpenDay,
    averageRating,
    reviewsCount,
    categories,
    operationalDays,
  ];
}
